File: backend/src/sandbox/isolated_workspace/network.py
"""Bridge + veth + nftables + IP pool for isolated workspaces (Linux only).

State at daemon scope:
  - One bridge ``eos-shared0`` with gateway ``10.244.0.1/24``.
  - One MASQUERADE rule on outbound from ``10.244.0.0/24``.
  - One IMDS drop rule (``169.254.169.254``) on the forward chain.
  - (Opt-in) RFC1918-deny drop rule when ``rfc1918_egress == "deny"``.

Per-workspace state:
  - One veth pair: ``eos-iws-{handle_id[:6]}h`` (host end on bridge) and
    ``eos-iws-{handle_id[:6]}n`` (peer end moved into the netns).
  - One ``/32`` allocation from ``10.244.0.2 - 10.244.0.254``.

The IP pool itself is pure Python; ``ip`` / ``nft`` calls are Linux-only and
raise ``IsolatedNetworkUnavailable`` if the tools are missing.
"""

from __future__ import annotations

import ipaddress
import shutil
import subprocess
from dataclasses import dataclass
from typing import Literal


BRIDGE_NAME = "eos-shared0"
BRIDGE_CIDR = ipaddress.IPv4Network("10.244.0.0/24")
GATEWAY = ipaddress.IPv4Address("10.244.0.1")
NFT_NAT_TABLE = "eos_iws_nat"
NFT_FILTER_TABLE = "eos_iws_filter"
IMDS_ADDR = "169.254.169.254"
RFC1918_NETS = ("10.0.0.0/8", "172.16.0.0/12", "192.168.0.0/16")
VETH_PREFIX = "eos-iws-"


class IsolatedNetworkUnavailable(RuntimeError):
    """Linux network primitives (ip / nft / CAP_NET_ADMIN) are not available."""


@dataclass(frozen=True)
class VethAllocation:
    host_name: str
    ns_ip: ipaddress.IPv4Address


class BridgeAddressPool:
    """Allocates /32s from ``10.244.0.2 - 10.244.0.254``.

    Pure Python, no Linux deps. Lowest-IP-first O(N) scan; N ≤ 253 so this
    is fine.
    """

    def __init__(self, network: ipaddress.IPv4Network = BRIDGE_CIDR) -> None:
        first = int(network.network_address) + 2  # skip .0 (network) and .1 (gw)
        last = int(network.broadcast_address) - 1  # skip .255 (broadcast)
        self._range = range(first, last + 1)
        self._allocated: set[ipaddress.IPv4Address] = set()

    def reserve(self, ip: ipaddress.IPv4Address) -> None:
        """Mark ``ip`` as in-use (used to rebuild pool state from manager.json)."""
        if int(ip) not in self._range:
            raise ValueError(f"{ip} is outside the pool range")
        self._allocated.add(ip)

    def allocate(self) -> ipaddress.IPv4Address:
        for value in self._range:
            candidate = ipaddress.IPv4Address(value)
            if candidate not in self._allocated:
                self._allocated.add(candidate)
                return candidate
        raise IsolatedNetworkUnavailable("isolated_workspace_ip_pool_exhausted")

    def free(self, ip: ipaddress.IPv4Address) -> None:
        self._allocated.discard(ip)


class IsolatedNetwork:
    """Owns ``eos-shared0`` bridge + static nft rules + per-ws veth wiring."""

    def __init__(
        self,
        *,
        rfc1918_egress: Literal["allow", "deny"] = "allow",
        pool: BridgeAddressPool | None = None,
    ) -> None:
        self.rfc1918_egress = rfc1918_egress
        self.pool = pool or BridgeAddressPool()
        self._initialized = False

    @property
    def initialized(self) -> bool:
        return self._initialized

    def initialize(self) -> None:
        """Install bridge + MASQUERADE + IMDS drop. Idempotent."""
        self._require_tools()
        self._ensure_bridge()
        self._install_static_rules()
        self._initialized = True

    def install_veth(self, *, handle_id: str, root_pid: int) -> VethAllocation:
        """Create veth pair, attach host end to bridge with port isolation.

        Also configures the ns-side end inside the iws's net namespace:
        brings it up, assigns ``ns_ip`` with the /24 prefix, and adds a
        default route via the bridge gateway. Without this the iws's veth
        is a bare interface — no IP, no link up, no route — and outbound
        traffic (ping/curl/DNS) fails with "Network unreachable". The
        ns-side config is intentionally minimal (single /32 in the bridge
        /24, default via gateway) so the design's MASQUERADE postrouting
        rule is the sole egress path.
        """
        if not self._initialized:
            raise IsolatedNetworkUnavailable("isolated_network_not_initialized")
        host, ns = _veth_names(handle_id)
        ns_ip = self.pool.allocate()
        try:
            _ip("link", "add", host, "type", "veth", "peer", "name", ns)
            _ip("link", "set", ns, "netns", str(root_pid))
            _ip("link", "set", host, "master", BRIDGE_NAME)
            _ip(
                "link",
                "set",
                host,
                "type",
                "bridge_slave",
                "isolated",
                "on",
                "mcast_flood",
                "off",
            )
            _ip("link", "set", host, "up")
            # Configure the ns-side inside the iws net namespace via nsenter.
            # The iws-side ns_holder.py is the holder of the new net ns but it
            # only brings up ``lo`` and purges IPv6; veth-side IP+route are
            # left to the daemon (this code) so the ns_holder stays minimal.
            prefix = BRIDGE_CIDR.prefixlen
            _ip_ns(root_pid, "link", "set", ns, "up")
            _ip_ns(root_pid, "addr", "add", f"{ns_ip}/{prefix}", "dev", ns)
            _ip_ns(root_pid, "route", "add", "default", "via", str(GATEWAY))
        except Exception:
            self.pool.free(ns_ip)
            _ip_quiet("link", "del", host)
            raise
        return VethAllocation(host_name=host, ns_ip=ns_ip)

    def teardown_veth(self, allocation: VethAllocation) -> None:
        _ip_quiet("link", "del", allocation.host_name)
        self.pool.free(allocation.ns_ip)

    def daemon_private_routes(self) -> list[str]:
        """Best-effort enumerate private routes visible from the daemon's netns.

        Surfaces the daemon-host pivot surface (plan §4 step 7 / Scenario 5).
        """
        try:
            result = subprocess.run(
                ["ip", "-o", "route", "show"],
                capture_output=True,
                text=True,
                check=False,
            )
        except FileNotFoundError:
            return []
        routes: list[str] = []
        for line in result.stdout.splitlines():
            dst = line.split(maxsplit=1)[0] if line else ""
            if _is_host_private_route(dst):
                routes.append(dst)
        return routes

    # ------------------------------------------------------------------
    def _ensure_bridge(self) -> None:
        existing = subprocess.run(
            ["ip", "-o", "link", "show", "dev", BRIDGE_NAME],
            capture_output=True,
            text=True,
            check=False,
        )
        if existing.returncode != 0:
            _ip("link", "add", BRIDGE_NAME, "type", "bridge")
            _ip("addr", "add", f"{GATEWAY}/24", "dev", BRIDGE_NAME)
            _ip("link", "set", BRIDGE_NAME, "up")

    def _install_static_rules(self) -> None:
        _nft("add", "table", "inet", NFT_NAT_TABLE)
        _nft(
            "add",
            "chain",
            "inet",
            NFT_NAT_TABLE,
            "postrouting",
            "{ type nat hook postrouting priority 100 ; }",
        )
        _nft(
            "add",
            "rule",
            "inet",
            NFT_NAT_TABLE,
            "postrouting",
            f'ip saddr {BRIDGE_CIDR} oifname != "{BRIDGE_NAME}" masquerade',
        )
        _nft("add", "table", "inet", NFT_FILTER_TABLE)
        _nft(
            "add",
            "chain",
            "inet",
            NFT_FILTER_TABLE,
            "forward",
            "{ type filter hook forward priority 0 ; }",
        )
        _nft("add", "rule", "inet", NFT_FILTER_TABLE, "forward", f"ip daddr {IMDS_ADDR} drop")
        if self.rfc1918_egress == "deny":
            for net in RFC1918_NETS:
                _nft(
                    "add",
                    "rule",
                    "inet",
                    NFT_FILTER_TABLE,
                    "forward",
                    f"ip daddr {net} ip daddr != {BRIDGE_CIDR} drop",
                )

    def _require_tools(self) -> None:
        for tool in ("ip", "nft"):
            if shutil.which(tool) is None:
                raise IsolatedNetworkUnavailable(f"missing required tool: {tool}")


def _veth_names(handle_id: str) -> tuple[str, str]:
    # Linux IFNAMSIZ caps interface names at 15 chars.
    # VETH_PREFIX (8: "eos-iws-") + short (6) + suffix (1) = 15 exactly.
    short = handle_id[:6]
    return f"{VETH_PREFIX}{short}h", f"{VETH_PREFIX}{short}n"


def _is_host_private_route(destination: str) -> bool:
    try:
        route = ipaddress.ip_network(destination, strict=False)
    except ValueError:
        return False
    if not isinstance(route, ipaddress.IPv4Network):
        return False
    if route.subnet_of(BRIDGE_CIDR):
        return False
    return any(route.overlaps(ipaddress.IPv4Network(net)) for net in RFC1918_NETS)


def _ip(*args: str) -> None:
    result = subprocess.run(
        ["ip", *args],
        capture_output=True,
        text=True,
        check=False,
    )
    if result.returncode != 0:
        # Surface stderr in the exception so debugging cascading network
        # failures doesn't require re-running with a custom logger. The
        # default ``check=True`` swallows it and only shows the cmd + rc.
        raise RuntimeError(
            f"ip {' '.join(args)} failed rc={result.returncode}: stderr={result.stderr.strip()!r}"
        )


def _ip_quiet(*args: str) -> None:
    subprocess.run(
        ["ip", *args],
        check=False,
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )


def _ip_ns(root_pid: int, *args: str) -> None:
    """Run ``ip <args>`` inside the net namespace owned by ``root_pid``.

    Uses nsenter against ``/proc/<pid>/ns/net`` rather than ``setns(2)`` from
    Python so the call stays single-step and inherits the daemon's existing
    capabilities. The iws ns_holder's user_ns is the parent's (mapped via
    --map-root-user), so root in the daemon's user_ns is root inside the
    iws net_ns — no extra capability dance needed for ip address/route ops.
    """
    subprocess.run(
        ["nsenter", "-t", str(root_pid), "-n", "--", "ip", *args],
        check=True,
        capture_output=True,
        text=True,
    )


def _nft(*args: str) -> None:
    """Run `nft` and ignore EEXIST (`File exists`) so initialize() stays idempotent."""
    result = subprocess.run(
        ["nft", *args],
        capture_output=True,
        text=True,
        check=False,
    )
    if result.returncode == 0:
        return
    err = result.stderr.lower()
    if "file exists" in err or "already exists" in err:
        return
    raise subprocess.CalledProcessError(
        result.returncode,
        result.args,
        output=result.stdout,
        stderr=result.stderr,
    )

File: sandbox/crates/eos-workspace-runtime/src/isolated/network.rs
//! Shared bridge + per-workspace veth + nftables wiring for isolated workspaces.
//!
//! Daemon-scope state: one bridge `eos-shared0` with gateway `10.244.0.1/24`, a
//! MASQUERADE rule on outbound from `10.244.0.0/24`, an IMDS drop rule, and an
//! opt-in RFC1918-deny rule. Per-workspace state: one veth pair and one `/32`
//! from `10.244.0.2 - 10.244.0.254`.
//!
//!
//! # IPv6 hardening — shell-free port
//!
//! The Rust holder shells out (`sysctl`, `ip -6 route flush`, `ip link set lo
//! up`) to purge IPv6 default routes and disable router-advertisement
//! acceptance so the v4-only MASQUERADE rule stays the sole egress. The Rust
//! port replaces those binaries with `rtnetlink` (`RTM_DELROUTE` for the IPv6
//! default route, `RTM_NEWLINK` to bring `lo` up) and direct
//! `/proc/sys/net/ipv6/conf/<iface>/accept_ra` writes — NO `ip`/`sysctl`
//! binaries. This work executes inside the namespace via `eos-ns-holder`
//! (see `host_runtime`), not in this daemon-scope module.

use std::net::Ipv4Addr;

use crate::isolated::caps::{Rfc1918Egress, HANDLE_PREFIX};
use crate::isolated::error::IsolatedError;

#[cfg(target_os = "linux")]
mod netfilter;
#[cfg(target_os = "linux")]
mod rtnl;

#[cfg(target_os = "linux")]
use netfilter::install_static_rules;
#[cfg(target_os = "linux")]
use rtnl::{ensure_bridge, ignore_not_found, install_veth_pair, link_index, run_netlink};

/// Shared bridge interface name.
pub const BRIDGE_NAME: &str = "eos-shared0";
/// Shared bridge CIDR.
pub(crate) const BRIDGE_CIDR: &str = "10.244.0.0/24";
/// Bridge gateway address.
pub const GATEWAY: &str = "10.244.0.1";
/// nftables NAT table name.
pub const NFT_NAT_TABLE: &str = "eos_iws_nat";
/// nftables filter table name.
pub const NFT_FILTER_TABLE: &str = "eos_iws_filter";
/// Cloud IMDS address dropped on the forward chain.
pub const IMDS_ADDR: &str = "169.254.169.254";
/// RFC1918 private networks (for the opt-in deny rule).
pub const RFC1918_NETS: [&str; 3] = ["10.0.0.0/8", "172.16.0.0/12", "192.168.0.0/16"];
/// Per-workspace veth name prefix — the SAME literal as [`HANDLE_PREFIX`].
pub(crate) const VETH_PREFIX: &str = HANDLE_PREFIX;

/// Bridge CIDR prefix length (matches `BRIDGE_CIDR`).
pub const BRIDGE_PREFIX_LEN: u8 = 24;
/// First allocatable host octet (skips `.0` network + `.1` gateway).
pub(crate) const POOL_FIRST_HOST: u8 = 2;
/// Last allocatable host octet (skips `.255` broadcast).
pub(crate) const POOL_LAST_HOST: u8 = 254;

/// One veth `/32` allocation for a workspace.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct VethAllocation {
    /// Host-side veth name (attached to the bridge).
    pub host_name: String,
    /// Namespace-side veth name (moved into the holder netns).
    pub ns_name: String,
    /// Namespace-side IPv4 address allocated from the pool.
    pub ns_ip: Ipv4Addr,
}

/// Host/peer veth names for a workspace handle.
///
/// Linux `IFNAMSIZ` caps names at 15 chars: `eos-iws-` (8) + 6 handle chars +
/// suffix (1) = 15 exactly. Host ends in `h`, peer ends in `n`.
#[must_use]
pub(crate) fn veth_names(workspace_handle_id: &str) -> (String, String) {
    // Rust `_veth_names` takes the FIRST 6 chars (`workspace_handle_id[:6]`);
    // reproduce that exactly so the interface naming is parity-equal.
    let short: String = workspace_handle_id.chars().take(6).collect();
    (
        format!("{VETH_PREFIX}{short}h"),
        format!("{VETH_PREFIX}{short}n"),
    )
}

/// Pure IPv4 `/32` allocator over `10.244.0.2 - 10.244.0.254`.
///
/// Lowest-IP-first O(N) scan; N <= 253. No Linux deps.
#[derive(Debug, Clone, Default)]
pub(crate) struct BridgeAddressPool {
    allocated: Vec<Ipv4Addr>,
}

impl BridgeAddressPool {
    /// Build an empty pool spanning the bridge CIDR's allocatable range.
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// Mark `ip` as in-use (used to rebuild pool state from `manager.json`).
    ///
    /// # Errors
    ///
    /// Returns [`IsolatedError::NetworkUnavailable`] when `ip` is outside
    /// [`BRIDGE_CIDR`].
    pub fn reserve(&mut self, ip: Ipv4Addr) -> Result<(), IsolatedError> {
        if !is_pool_ip(ip) {
            return Err(IsolatedError::NetworkUnavailable(format!(
                "isolated workspace IP {ip} is outside {BRIDGE_CIDR}"
            )));
        }
        if !self.allocated.contains(&ip) {
            self.allocated.push(ip);
            self.allocated.sort_unstable();
        }
        Ok(())
    }

    /// Allocate the lowest free `/32` in the pool.
    ///
    /// # Errors
    ///
    /// Returns [`IsolatedError::NetworkUnavailable`] when the address pool is
    /// exhausted.
    pub fn allocate(&mut self) -> Result<Ipv4Addr, IsolatedError> {
        for host in POOL_FIRST_HOST..=POOL_LAST_HOST {
            let ip = Ipv4Addr::new(10, 244, 0, host);
            if !self.allocated.contains(&ip) {
                self.allocated.push(ip);
                self.allocated.sort_unstable();
                return Ok(ip);
            }
        }
        Err(IsolatedError::NetworkUnavailable(
            "isolated_workspace_ip_pool_exhausted".to_owned(),
        ))
    }

    /// Release `ip` back into the pool.
    pub fn free(&mut self, ip: Ipv4Addr) {
        self.allocated.retain(|allocated| *allocated != ip);
    }
}

/// Owns the `eos-shared0` bridge + static nft rules + per-workspace veth wiring.
///
/// The Rust implementation shells out to `ip`/`nft`; the Rust port replaces
/// the bridge/veth path with `rtnetlink` link/address operations and the static
/// NAT/filter path with `NETLINK_NETFILTER` messages — NO `ip`/`nft` binaries.
#[derive(Debug)]
pub(crate) struct IsolatedNetwork {
    rfc1918_egress: Rfc1918Egress,
    pool: BridgeAddressPool,
    initialized: bool,
}

impl IsolatedNetwork {
    /// Construct an uninitialized network with the given egress policy.
    #[must_use]
    pub fn new(rfc1918_egress: Rfc1918Egress) -> Self {
        Self {
            rfc1918_egress,
            pool: BridgeAddressPool::new(),
            initialized: false,
        }
    }

    /// Install the bridge + MASQUERADE + IMDS drop (+ optional RFC1918 deny).
    /// Idempotent for table/chain creation; rule insertion mirrors Rust's
    /// sequential `nft add rule` calls.
    ///
    /// # Errors
    ///
    /// Returns [`IsolatedError::NetworkUnavailable`] when route or netfilter
    /// netlink setup fails.
    pub fn initialize(&mut self) -> Result<(), IsolatedError> {
        if test_harness_enabled() {
            self.initialized = true;
            return Ok(());
        }
        let rfc1918_egress = self.rfc1918_egress;
        #[cfg(target_os = "linux")]
        {
            run_netlink(move |handle| async move {
                let bridge_index = ensure_bridge(&handle).await?;
                install_static_rules(rfc1918_egress, bridge_index)?;
                Ok(())
            })?;
        }
        #[cfg(not(target_os = "linux"))]
        {
            let _ = rfc1918_egress;
        }
        self.initialized = true;
        Ok(())
    }

    /// Create a veth pair, attach the host end to the bridge with port
    /// isolation, and configure the namespace-side end (up, `/24` addr,
    /// default route via gateway).
    ///
    /// # Errors
    ///
    /// Returns [`IsolatedError::NetworkUnavailable`] when bridge initialization,
    /// IP allocation, holder PID conversion, or veth netlink wiring fails.
    pub fn install_veth(
        &mut self,
        workspace_handle_id: &str,
        holder_pid: i32,
    ) -> Result<VethAllocation, IsolatedError> {
        if !self.initialized {
            self.initialize()?;
        }
        let (host_name, ns_name) = veth_names(workspace_handle_id);
        let ns_ip = self.pool.allocate()?;
        if !test_harness_enabled() && holder_pid > 0 {
            #[cfg(target_os = "linux")]
            {
                let host = host_name.clone();
                let ns = ns_name.clone();
                let holder_pid = u32::try_from(holder_pid).map_err(|_| {
                    IsolatedError::NetworkUnavailable(format!(
                        "invalid isolated holder pid {holder_pid}"
                    ))
                })?;
                if let Err(error) = run_netlink(move |handle| async move {
                    install_veth_pair(&handle, &host, &ns, holder_pid).await
                }) {
                    self.pool.free(ns_ip);
                    return Err(error);
                }
            }
        }
        Ok(VethAllocation {
            host_name,
            ns_name,
            ns_ip,
        })
    }

    /// Tear down a veth pair and return its `/32` to the pool.
    pub fn teardown_veth(&mut self, allocation: &VethAllocation) {
        self.teardown_host_veth(&allocation.host_name);
        self.pool.free(allocation.ns_ip);
    }

    /// Delete a host-side veth by name without touching the IP pool.
    ///
    /// Startup/test cleanup uses this for naming-convention orphan sweeps where
    /// no persisted `/32` allocation is available.
    pub fn teardown_host_veth(&mut self, host_name: &str) {
        #[cfg(not(target_os = "linux"))]
        let _ = host_name;
        if !test_harness_enabled() {
            #[cfg(target_os = "linux")]
            {
                let host_name = host_name.to_owned();
                let _ = run_netlink(move |handle| async move {
                    if let Some(index) = link_index(&handle, &host_name).await? {
                        ignore_not_found(
                            "delete host veth",
                            handle.link().del(index).execute().await,
                        )?;
                    }
                    Ok(())
                });
            }
        }
    }

    /// Reserve a previously persisted namespace IP during startup reconciliation.
    ///
    /// Crash-recovered handles are reaped before new enters run, but their
    /// persisted IPs stay reserved for this daemon lifetime so a fresh handle
    /// cannot race a stale kernel resource that is still being torn down.
    ///
    /// # Errors
    ///
    /// Returns [`IsolatedError::NetworkUnavailable`] when `ip` is outside the
    /// isolated bridge pool.
    pub fn reserve_persisted_ip(&mut self, ip: Ipv4Addr) -> Result<(), IsolatedError> {
        self.pool.reserve(ip)
    }
}

fn test_harness_enabled() -> bool {
    std::env::var("EOS_ISOLATED_WORKSPACE_TEST_HARNESS")
        .is_ok_and(|value| matches!(value.as_str(), "1" | "true" | "TRUE" | "yes" | "YES"))
}

/// Wrap a netlink/netfilter failure as [`IsolatedError::NetworkUnavailable`],
/// prefixing the failing step. Shared by the rtnetlink and nft wire paths.
#[cfg(target_os = "linux")]
pub(crate) fn network_error_at(
    step: impl Into<String>,
    error: impl std::fmt::Display,
) -> IsolatedError {
    IsolatedError::NetworkUnavailable(format!("{}: {error}", step.into()))
}

fn is_pool_ip(ip: Ipv4Addr) -> bool {
    let octets = ip.octets();
    octets[0] == 10
        && octets[1] == 244
        && octets[2] == 0
        && (POOL_FIRST_HOST..=POOL_LAST_HOST).contains(&octets[3])
}
